File: app/src/main/java/doit/study/droid/activities/InterrogatorActivity.java
package doit.study.droid.activities;

import android.database.Cursor;
import android.net.Uri;
import android.os.Bundle;
import android.os.Handler;
import android.support.v4.app.LoaderManager;
import android.support.v4.content.CursorLoader;
import android.support.v4.content.Loader;
import android.support.v4.view.PagerTabStrip;
import android.support.v4.view.ViewPager;
import android.view.View;
import android.widget.TextView;

import doit.study.droid.R;
import doit.study.droid.adapters.InterrogatorPagerAdapter;
import doit.study.droid.data.Question;
import doit.study.droid.data.QuizProvider;
import doit.study.droid.data.RelationTables;
import doit.study.droid.fragments.InterrogatorFragment;
import timber.log.Timber;


public class InterrogatorActivity extends DrawerBaseActivity implements InterrogatorFragment.OnFragmentActivityChatter, LoaderManager.LoaderCallbacks<Cursor> {
    private static final boolean DEBUG = false;
    private ViewPager mPager;
    // counters for current test
    private static final String WRONG_CNT_KEY = "doit.study.dodroid.wrong_cnt_key";
    private static final String RIGHT_CNT_KEY = "doit.study.dodroid.right_cnt_key";
    private int mRightCnt;
    private int mWrongCnt;
    private static final String PAGE_INDEX_IN_FOCUS_KEY = "doit.study.dodroid.page_index_in_focus_key";
    private int mCurrentPageInFocus;
    private final int QUIZ_SIZE = 10; // default quiz size
    private int mQuizSize; // actual size can be lesser
    private static final String PROGRESS_KEY = "doit.study.dodroid.progress_key";
    private int mProgress = -1; // quantity of answered questions
    private static final int QUESTION_LOADER = 0;
    private InterrogatorPagerAdapter mPagerAdapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (DEBUG) Timber.d("onCreate");
        mSelectionId = R.id.nav_do_it;
        getLayoutInflater().inflate(R.layout.activity_interrogator, mContainerContent);
        getSupportLoaderManager().initLoader(QUESTION_LOADER, null, InterrogatorActivity.this);
        mPager = (ViewPager) findViewById(R.id.view_pager);
        configPagerTabStrip();
        mPagerAdapter = new InterrogatorPagerAdapter(getSupportFragmentManager(), this);
        mPager.setAdapter(mPagerAdapter);

        if (savedInstanceState != null){
            if (DEBUG) Timber.d("Restore saved state");
            mRightCnt = savedInstanceState.getInt(RIGHT_CNT_KEY);
            mWrongCnt = savedInstanceState.getInt(WRONG_CNT_KEY);
            mProgress = savedInstanceState.getInt(PROGRESS_KEY);
            if (mProgress == 0) {
                showProgress();
            } else {
                mCurrentPageInFocus = savedInstanceState.getInt(PAGE_INDEX_IN_FOCUS_KEY);
                if (DEBUG) Timber.d("Cur page: %d", mCurrentPageInFocus);
                mPager.setCurrentItem(mCurrentPageInFocus, true);
            }
        }
        setTitle(getResources().getQuantityString(R.plurals.numberOfQuestionsInTest, mProgress, mProgress));
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        if (DEBUG) Timber.d("onSaveInstanceState");
        outState.putInt(WRONG_CNT_KEY, mWrongCnt);
        outState.putInt(RIGHT_CNT_KEY, mRightCnt);
        outState.putInt(PROGRESS_KEY, mProgress);
        outState.putInt(PAGE_INDEX_IN_FOCUS_KEY, mPager.getCurrentItem());
        super.onSaveInstanceState(outState);
    }


    private void configPagerTabStrip() {
        PagerTabStrip pagerTabStrip = (PagerTabStrip) mPager.findViewById(R.id.pager_title_strip);
        // show one title
        pagerTabStrip.setNonPrimaryAlpha(0);
        // set the black underlining
        pagerTabStrip.setTabIndicatorColor(0x000000);
    }


    @Override
    public void saveStat(Question question) {
        if (DEBUG) Timber.d("saveStat %s", question);
        getContentResolver().update(QuizProvider.QUESTION_URI, Question.getContentValues(question), "_ID = " + question.getId(), null);
    }

    @Override
    public void updateProgress(boolean isRight) {
        if (isRight)
            ++mRightCnt;
        else
            ++mWrongCnt;
        --mProgress;
        setTitle(getResources().getQuantityString(R.plurals.numberOfQuestionsInTest, mProgress, mProgress));
        if (mProgress == 0) {
            showProgress();
        }
    }

    private void showProgress(){
        Handler handler = new Handler();
        handler.postDelayed(new Runnable() {
            public void run() {
                if (DEBUG) Timber.d("swipe to the result page");
                mPagerAdapter.addResultPage(mRightCnt, mWrongCnt);
                setTitle(getString(R.string.test_completed));
                mPager.setCurrentItem(mQuizSize, true);
            }
        }, 2000);
    }

    @Override
    public void swipeToNext(int delay) {
        final int posInFocus = mPager.getCurrentItem();
        Handler handler = new Handler();
        handler.postDelayed(new Runnable() {
            @Override
            public void run() {
                if (posInFocus == mPager.getCurrentItem()) {
                    if (DEBUG) Timber.d("swipe to the next page");
                    mPager.setCurrentItem(posInFocus + 1, true);
                }
            }
        }, delay);
    }

    @Override
    public Loader<Cursor> onCreateLoader(int id, Bundle args) {
        Uri randQuestionUri = QuizProvider.QUESTION_URI
                .buildUpon()
                .appendPath("rand").appendPath(Integer.toString(QUIZ_SIZE))
                .build();
        return new CursorLoader(this, randQuestionUri, RelationTables.JoinedQuestionTagProjection, null, null, null);
    }

    @Override
    public void onLoadFinished(Loader<Cursor> loader, Cursor data) {
        if (DEBUG) Timber.d("load finished: %d", data.hashCode());
        int size = data.getCount();
        if (size == 0) {
            TextView noTopic = (TextView) findViewById(R.id.no_topic_selected);
            noTopic.setVisibility(View.VISIBLE);
            mPager.setVisibility(View.GONE);
        }
        // load just once
        else if (mQuizSize == 0) {
            mQuizSize = size;
            if (mProgress == -1)
                mProgress = size;
            mPagerAdapter.setData(data);
            mPager.setCurrentItem(mCurrentPageInFocus, true);
            setTitle(getResources().getQuantityString(R.plurals.numberOfQuestionsInTest, mProgress, mProgress));
        }
    }

    @Override
    public void onLoaderReset(Loader<Cursor> loader) {
        if (DEBUG) Timber.d("onLoaderReset");
    }
}

File: app/src/main/java/doit/study/droid/activities/InterrogatorActivity.kt
package doit.study.droid.activities

import android.database.Cursor
import android.os.Bundle
import android.os.Handler
import android.view.View
import android.widget.TextView


import androidx.loader.app.LoaderManager
import androidx.loader.content.CursorLoader
import androidx.loader.content.Loader
import androidx.viewpager.widget.PagerTabStrip
import androidx.viewpager.widget.ViewPager

import doit.study.droid.R
import doit.study.droid.adapters.InterrogatorPagerAdapter
import doit.study.droid.data.Question
import doit.study.droid.data.QuizProvider
import doit.study.droid.data.RelationTables
import doit.study.droid.fragments.InterrogatorFragment
import timber.log.Timber


class InterrogatorActivity : DrawerBaseActivity(), InterrogatorFragment.OnFragmentActivityChatter, LoaderManager.LoaderCallbacks<Cursor> {
    private var pager: ViewPager? = null
    private var rightCnt: Int = 0
    private var wrongCnt: Int = 0
    private var currentPageInFocus: Int = 0
    private var quizSize: Int = 0 // actual size can be lesser
    private var aprogress = -1 // quantity of answered questions
    private var pagerAdapter: InterrogatorPagerAdapter? = null
    private var handler: Handler? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (DEBUG) Timber.d("onCreate")
        selectionId = R.id.nav_do_it

        handler = Handler()

        layoutInflater.inflate(R.layout.activity_interrogator, containerContent)
        supportLoaderManager.initLoader(QUESTION_LOADER, null, this@InterrogatorActivity)
        pager = findViewById(R.id.view_pager)
        configPagerTabStrip()
        pagerAdapter = InterrogatorPagerAdapter(supportFragmentManager, this)
        pager?.adapter = pagerAdapter

        savedInstanceState?.let {
            if (DEBUG) Timber.d("Restore saved state")
            rightCnt = it.getInt(RIGHT_CNT_KEY)
            wrongCnt = it.getInt(WRONG_CNT_KEY)
            aprogress = it.getInt(PROGRESS_KEY)
            if (aprogress == 0) {
                showProgress()
            } else {
                currentPageInFocus = it.getInt(PAGE_INDEX_IN_FOCUS_KEY)
                if (DEBUG) Timber.d("Cur page: %d", currentPageInFocus)
                pager?.setCurrentItem(currentPageInFocus, true)
            }
        }
        val questionsLeft = if (aprogress >= 0) aprogress else 0 // initial value -1
        title = resources.getQuantityString(R.plurals.numberOfQuestionsInTest, questionsLeft, questionsLeft)
    }

    override fun onSaveInstanceState(outState: Bundle) {
        if (DEBUG) Timber.d("onSaveInstanceState")
        outState.putInt(WRONG_CNT_KEY, wrongCnt)
        outState.putInt(RIGHT_CNT_KEY, rightCnt)
        outState.putInt(PROGRESS_KEY, aprogress)
        outState.putInt(PAGE_INDEX_IN_FOCUS_KEY, pager!!.currentItem)
        super.onSaveInstanceState(outState)
    }


    private fun configPagerTabStrip() {
        val pagerTabStrip = pager!!.findViewById<PagerTabStrip>(R.id.pager_title_strip)
        // show one title
        pagerTabStrip.setNonPrimaryAlpha(0f)
        // set the black underlining
        pagerTabStrip.tabIndicatorColor = 0x000000
    }


    override fun saveStat(question: Question) {
        if (DEBUG) Timber.d("saveStat %s", question)
        contentResolver.update(QuizProvider.QUESTION_URI, Question.getContentValues(question), "_ID = " + question.id, null)
    }

    override fun updateProgress(isRight: Boolean) {
        if (isRight)
            ++rightCnt
        else
            ++wrongCnt
        --aprogress
        title = resources.getQuantityString(R.plurals.numberOfQuestionsInTest, aprogress, aprogress)
        if (aprogress == 0) {
            showProgress()
        }
    }

    private fun showProgress() {
        handler?.postDelayed({
            if (DEBUG) Timber.d("swipe to the result page")
            pagerAdapter?.addResultPage(rightCnt, wrongCnt)
            title = getString(R.string.test_completed)
            pager?.setCurrentItem(quizSize, true)
        }, 2000)
    }

    override fun onDestroy() {
        super.onDestroy()
        handler?.removeCallbacksAndMessages(null)
    }

    override fun swipeToNext(delay: Int) {
        val posInFocus = pager!!.currentItem
        val handler = Handler()
        handler.postDelayed({
            if (posInFocus == pager!!.currentItem) {
                if (DEBUG) Timber.d("swipe to the next page")
                pager!!.setCurrentItem(posInFocus + 1, true)
            }
        }, delay.toLong())
    }

    override fun onCreateLoader(id: Int, args: Bundle?): Loader<Cursor> {
        val randQuestionUri = QuizProvider.QUESTION_URI
                .buildUpon()
                .appendPath("rand").appendPath(Integer.toString(QUIZ_SIZE))
                .build()
        return CursorLoader(this, randQuestionUri, RelationTables.JoinedQuestionTagProjection, null, null, null)
    }

    override fun onLoadFinished(loader: Loader<Cursor>, data: Cursor) {
        if (DEBUG) Timber.d("load finished: %d", data.hashCode())
        val size = data.count
        if (size == 0) {
            val noTopic = findViewById<View>(R.id.no_topic_selected) as TextView
            noTopic.visibility = View.VISIBLE
            pager!!.visibility = View.GONE
        } else if (quizSize == 0) {
            quizSize = size
            if (aprogress == -1)
                aprogress = size
            pagerAdapter!!.setData(data)
            pager!!.setCurrentItem(currentPageInFocus, true)
            title = resources.getQuantityString(R.plurals.numberOfQuestionsInTest, aprogress, aprogress)
        }// load just once
    }

    override fun onLoaderReset(loader: Loader<Cursor>) {
        if (DEBUG) Timber.d("onLoaderReset")
    }

    companion object {
        private const val DEBUG = false
        // counters for current test
        private const val WRONG_CNT_KEY = "doit.study.dodroid.wrong_cnt_key"
        private const val RIGHT_CNT_KEY = "doit.study.dodroid.right_cnt_key"
        private const val PAGE_INDEX_IN_FOCUS_KEY = "doit.study.dodroid.page_index_in_focus_key"
        private const val PROGRESS_KEY = "doit.study.dodroid.progress_key"
        private const val QUESTION_LOADER = 0
        private const val QUIZ_SIZE = 10 // default quiz size
    }
}

